### PR TITLE
Add an integration test library and some integration tests for scheduler

### PR DESCRIPTION
**Pull request by bsalamat**
**What this PR does / why we need it**:

1. Add an integration test library (utils.go) for scheduler testing.
2. Cleaned up some of the tests in scheduler_test.go with the new integration test library.
3. Add priority_test.go with a couple of examples on how to test scheduler priority function in integration tests.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

ref/ #48176

@kubernetes/sig-scheduling-pr-reviews 
@davidopp @k82cn @vikaschoudhary16

### CODE REVIEW

File: test/integration/scheduler/utils.go
Comment (davidopp) [inline test/integration/scheduler/utils.go]: first "created" should be "creates"

File: test/integration/scheduler/utils.go
Comment (davidopp) [inline test/integration/scheduler/utils.go]: s/return/returns/

File: test/integration/scheduler/utils.go
Comment (davidopp) [inline test/integration/scheduler/utils.go]: s/return/returns/

File: test/integration/scheduler/priorities_test.go
Comment (davidopp) [inline test/integration/scheduler/priorities_test.go]: Bless you, I think this may be the first file in the entire codebase that has a comment explaining the purpose of the file.

**Comment by davidopp** [comment]
I have a theory about why the pod affinity test is failing. All the pods in that test except the last one (the test one) have NodeName set. This means they don't go through the scheduler (scheduler only looks at pods with empty NodeName). My guess is that the scheduler hasn't seen the attractor pod running on the kubelet yet, at the time when you create the pod with pod affinity. (`runPausePod()` calls `waitForPodToSchedule()` which trivially returns true for any pod that had the nodeName explicitly set when it was created, so you don't really know whether the scheduler has seen any of the pods you created). 

I would suggest to either add some kind of wall-clock time wait or explicitly peek into the scheduler state to wait until all the pods have been observed by the scheduler, before creating the last pod (the one with pod affinity).

**Comment by bsalamat** [comment]
@davidopp `runPausePod` which is used to create and run "attractorPod" waits for the pod to get scheduled. When I printed the pods processed by `CalculateInterPodAffinityPriority` function, that pod is there.

**Comment by bsalamat** [comment]
@davidopp I just deleted my previous comment. I said it too soon. Let me look into the issue. You are probably right.

**Comment by bsalamat** [comment]
@davidopp My mistake was that I was expecting the new pod to the be scheduled on the same node as the "attractor pod". Since all the nodes were in the same topology domain, the inter-pod affinity score of all them were the same and scheduler was picking one. I fixed the test by adding a few nodes out of the topology domain and by checking that the new pod is scheduled in the same topology domain as the "attractor pod".

**Comment by davidopp** [comment]
Oh right -- sorry I didn't think of that!! A simple change from your original version is to give each of the nodes a different value of the topology key. But what you described also makes sense.

**Comment by davidopp** [comment]
Let me know when you want me to re-review.

**Comment by bsalamat** [comment]
@davidopp The PR is ready. PTAL.

File: test/integration/scheduler/priorities_test.go
Comment (davidopp) [inline test/integration/scheduler/priorities_test.go]: Shouldn't this be S1? It seems that this test is passing only by luck. Multiple MatchExpressions are ANDed, so the overall pod affinity term should be false on all nodes.

Comment (bsalamat) [inline test/integration/scheduler/priorities_test.go]: There are two values in the array. The content of the first one which is `labelValue` is 'S1'.

Comment (davidopp) [inline test/integration/scheduler/priorities_test.go]: Oops! You're right.

**Comment by bsalamat** [comment]
Thanks, @davidopp.

/assign @foxish 
For approval

**Comment by vikaschoudhary16** [comment]
@bsalamat looks like rebase is needed.

**Comment by bsalamat** [comment]
@davidopp @foxish rebased the PR. Please talk a look and approve if possible.

**Review by resouer — COMMENTED** [review]
ping @bsalamat Please rebase :)

**Review by timothysc — CHANGES_REQUESTED** [review]
I haven't done a full review, but found some issues that gave me pause.

File: test/integration/scheduler/utils.go
Comment (bsalamat) [inline test/integration/scheduler/utils.go]: It is only there to get the "pause image". I could replicate the code from e2e framework and remove the dependency.

File: test/integration/scheduler/utils.go
Comment (bsalamat) [inline test/integration/scheduler/utils.go]: I think you are confusing this with our e2e tests. I don't remember that deleting the namespace has been a source of problem in integration tests. What caused trouble in e2e tests was not deletion of the namespace, it was the fact that many of our priority and predicate tests were failing due to inconsistencies in our test clusters. Those clusters are not predictable enough for these kinds of tests.

Comment (timothysc) [inline test/integration/scheduler/utils.go]: There have been a number of issues that existed now where lengthy namespace deletes have been the root of e2e flakeyness.  With integration tests we will now share a single cluster across tests, and any cruft we leave around could affect other tests.

File: test/integration/scheduler/utils.go
Comment (timothysc) [inline test/integration/scheduler/utils.go]: We probably should abstract to use a common test/utils library for both on that one.

**Comment by jayunit100** [comment]
cc @ravisantoshgudimetla

File: test/integration/scheduler/utils.go
Comment (bsalamat) [inline test/integration/scheduler/utils.go]: Yes, we may be able to do so, but I'd leave that to the test team. For now, I am going to bring those functions to the integration test framework and remove the dependency on e2e framework.

File: test/integration/scheduler/utils.go
Comment (bsalamat) [inline test/integration/scheduler/utils.go]: Yes, I agree with that, but what do you suggest to solve the problem? I think we should perform cleanup in our library as much as possible and make sure our tests do cleanup as well.

Comment (timothysc) [inline test/integration/scheduler/utils.go]: Append a UUID to the NS-name, reduces random collisions.. and double check to make certain NS-deletes are non-blocking calls.

Comment (bsalamat) [inline test/integration/scheduler/utils.go]: I like your idea about using UUID for namespaces, but I think NS deletion should be blocking. I don't know how our test framework works exactly, but we probably don't want to have a namespace from an older test to be present when we start the next test.

**Comment by bsalamat** [comment]
@timothysc Done. PTAL.

File: test/integration/framework/util.go
Comment (ravisantoshgudimetla) [inline test/integration/framework/util.go]: Seems this function is not exported anywhere, can we change it to getServerArchitecture?

Comment (bsalamat) [inline test/integration/framework/util.go]: This is a copy/paste from e2e framework. I think this may be a useful method for other tests in the future.

File: test/integration/scheduler/util.go
Comment (ravisantoshgudimetla) [inline test/integration/scheduler/util.go]: Is it worth  to delete the pod that we created in test?

Comment (bsalamat) [inline test/integration/scheduler/util.go]: We delete the namespace that we create for the test. When a namespace is deleted, everything under the NS is deleted, including pods. If a test creates a new namespace, the test must ensure that it cleans up the namespace at the end of the test.

Comment (ravisantoshgudimetla) [inline test/integration/scheduler/util.go]: Got you, so we are still go ahead with deletion of namespace, thanks :)

Comment (bsalamat) [inline test/integration/scheduler/util.go]: Yes, the namespace has to be deleted, otherwise it may interfere with the next tests.

File: test/integration/scheduler/util.go
Comment (ravisantoshgudimetla) [inline test/integration/scheduler/util.go]: Looking at https://github.com/kubernetes/kubernetes/blob/master/test/integration/framework/master_utils.go#L409, it seems we are doing anything there? Hope this is the fn we are using. Is this intentional?

File: test/integration/scheduler/util.go
Comment (ravisantoshgudimetla) [inline test/integration/scheduler/util.go]: nit: err:=waitForPodToSchedule because if we don't have err in above line 276, this would have failed.

File: test/integration/scheduler/util.go
Comment (ravisantoshgudimetla) [inline test/integration/scheduler/util.go]: Ditto.

File: test/integration/scheduler/util.go
Comment (bsalamat) [inline test/integration/scheduler/util.go]: It wouldn't be more or less the same. err is already defined in the code. So, I wouldn't worry about it.

**Comment by ravisantoshgudimetla** [comment]
@bsalamat Overall the changes look good. I am wondering, if we could potentially share some code across the scheduler_perf and scheduler folders in integration tests. For example,
-  https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/util.go does the same api server/scheduler setup and gives a function back to tear it down. 
- Similarly, the pod creation in the other PR that I am working on https://github.com/kubernetes/kubernetes/pull/46117/files#diff-3126019cd381f25186f4fa8c8da06ec6R234 relies on e2e framework. 

WDYT?

**Comment by bsalamat** [comment]
@ravisantoshgudimetla I like that idea. We may want to have one `scheduler` directory under `integration` and have `perf` and `algorithm` (or whatever name) directories under the `scheduler` directory. The scheduler directory can have a library that all of these use. That should be done in a separate PR though. I don't want to mix many things in a single PR.

**Comment by ravisantoshgudimetla** [comment]
@bsalamat Agreed. I don't want to block this PR either. Just wanted to know your thoughts on sharing the code.

File: test/integration/scheduler/priorities_test.go
Comment (wanghaoran1988) [inline test/integration/scheduler/priorities_test.go]: I'd prefer using some random string, if the node clean up failed and make the cluster messed up, with random string, we can make sure the tests keep going on.

File: test/integration/scheduler/util.go
Comment (wanghaoran1988) [inline test/integration/scheduler/util.go]: This name isn't super descriptive, we are actually waiting for the Node with specific key pass the specific check.

Comment (wanghaoran1988) [inline test/integration/scheduler/util.go]: Also I didn't see anywhere we use this util func in this pr.

File: test/integration/scheduler/priorities_test.go
Comment (bsalamat) [inline test/integration/scheduler/priorities_test.go]: That would make the test more resilient, but I guess the tests are not supposed to be resilient. Tests are preferred to fail fast when a PR breaks anything in the test or in the integration cluster.

File: test/integration/scheduler/util.go
Comment (bsalamat) [inline test/integration/scheduler/util.go]: This function is used in scheduler_test.go. I didn't write this function. I only moved it from that file to util.go.

File: test/integration/scheduler/priorities_test.go
Comment (wanghaoran1988) [inline test/integration/scheduler/priorities_test.go]: I am not sure how much this will be flaky, but using random string should be considered to reduce the flaky tests, anyway, not a big problem, and will not against if you think it's ok

Comment (bsalamat) [inline test/integration/scheduler/priorities_test.go]: Given that we delete all the nodes at the end of each test, I am not sure why a random string should reduce flakes.

File: test/integration/framework/util.go
Comment (timothysc) [inline test/integration/framework/util.go]: could you please add a //TODO comment about potential to be moved to another common test library.

File: test/integration/scheduler/util.go
Comment (timothysc) [inline test/integration/scheduler/util.go]: gr8 thx.

File: test/integration/scheduler/util.go
Comment (timothysc) [inline test/integration/scheduler/util.go]: nit: could probably collapse these two f(n) by just having a nil check on resource.

File: test/integration/scheduler/util.go
Comment (timothysc) [inline test/integration/scheduler/util.go]: same nit.

File: test/integration/framework/util.go
Comment (bsalamat) [inline test/integration/framework/util.go]: Sure. Done.

**Review by timothysc — CHANGES_REQUESTED** [review]
Minor comments... mostly //TODO and others optional.

**Comment by timothysc** [comment]
Could you add the standard scheduling OWNERS file to that directory as well, then lgtm.

**Comment by bsalamat** [comment]
@timothysc Thanks again for your review. I addressed the comments.
Regarding the OWNERS file, the directory already has one.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Comment by timothysc** [comment]
I'll knock on PR to update OWNERS for some areas.

**Pull request merged**
